### DISCUSSION TITLE
Certificate Validation When Using Elliptic Curves With Explicit Params

### DISCUSSION THREAD

**Discussion by Jafaral** [strongswan/strongswan#1796] [General]
OpenSSL doesn't allow certs with explicit curve in verification chain [1] for security reasons [2]. I am using OpenSSL plugin with Strongswan so I expected to get this behavior covered by the plugin. This doesn't seem to be the case though. 

If I use Strongswan's pki:
```
$ pki --verify --cacert x509ca/ca-explicit.crt --in x509/user.crt 
  using certificate "C=US, O=Router, CN=user"
  using trusted ca certificate "C=US, O=Router, CN=router.ca"
  reached self-signed root ca with a path length of 0
certificate trusted, lifetimes valid
```
when using openssl to verify the cert, it works as expected:
```
$ openssl verify -CAfile x509ca/ca-explicit.crt  x509/user.crt
C = US, O = Router, CN = router.ca
error 94 at 1 depth lookup: Certificate public key has explicit ECC parameters
error x509/user.crt: verification failed
```

1- Does Strongswan consider whether the curve is named or explicit when validating certs?
2- Is there a way to get the cert validation governed by OpenSSL plugin?

[1] https://github.com/openssl/openssl/pull/12683
[2] https://lightshipsec.com/explicitly-parameterized-ecdsa-x-509-certificates/

**Comment by tobiasbrunner** [comment]
> 1- Does Strongswan consider whether the curve is named or explicit when validating certs?

No, currently not. We could probably add a check when parsing EC keys and reject any that use explicit parameters.

That's easy with the _wolfssl_ plugin. WolfSSL actually already rejects such keys unless it is built with `WOLFSSL_NO_ASN_STRICT`. It's a bit trickier with the _openssl_ plugin as some functions have been deprecated with with OpenSSL 3.0 (and were only added relatively late with 1.1.1). I'm not entirely sure about the _botan_ plugin, but Botan has deprecated explicit curve parameters, so maybe in the future they'll reject them as well. **Edit:** I've [asked about this at the Botan repo](https://github.com/randombit/botan/discussions/3630), let's see if anybody replies.

I've pushed some changes to the _1796-ecdsa-explicit_ branch.

> 2- Is there a way to get the cert validation governed by OpenSSL plugin?

No.

**Comment by Jafaral** [comment]
Thank you @tobiasbrunner !

I have not tested the daemon yet. `pki` looks promising but with caveats. 
```
pki --verify --cacert x509ca/ca-ec-explicit.crt  --in x509/user.crt 
OpenSSL X.509 parsing failed
building CRED_CERTIFICATE - X509 failed, tried 5 builders
parsing CA certificate from 'x509ca/ca-ec-explicit.crt' failed
  using trusted certificate "C=US, O=Router, CN=user"
no issuer certificate found for "C=US, O=Router, CN=user"
  issuer is "C=US, O=Router, CN=router.ca explicit"
  reached end of incomplete trust chain for trusted certificate "C=US, O=Router, CN=user"
certificate trusted, lifetimes valid
```
However, if I don't explicitly pick the ca cert:

```
pki --verify --cacert x509ca   --in x509/user.crt 
OpenSSL X.509 parsing failed
building CRED_CERTIFICATE - X509 failed, tried 5 builders
parsing CA certificate from 'x509ca/ca-ec-explicit.crt' failed
  using certificate "C=US, O=Router, CN=user"
no issuer certificate found for "C=US, O=Router, CN=user"
  issuer is "C=US, O=Router, CN=router.ca explicit"
certificate untrusted
``` 
The second case reports  the expected `certificate untrusted` only because I have other CAs in the directory. If I delete those and only keep  `ca-ec-explicit.crt`, the output is the same as the first case `certificate trusted, lifetimes valid`. This looks like a bug, unless I'm missing something.

**Reply by tobiasbrunner** [reply]
The first case is basically as if not configuring a CA certificate as only successfully parsed certificates that are added to the trust store are considered by the `--verify` command. Without CA certificates, it adds the end-entity certificate to the trust store and verifies it as if it was configured as a local certificate (i.e. it still tries to build a trust chain, but it's fine if it's incomplete).

**Comment by Jafaral** [comment]
Other observations:

I have this config file:
```
authorities {
  expCA {
      cacert = /etc/swanctl/x509ca/ca-ec-explicit.crt
   }
}
connections {
   r3 {
      version = 2
      local_addrs = 192.168.229.110
      remote_addrs = %any
      local {
         auth = ecdsa
         certs = /etc/swanctl/x509/router.crt
      }
      remote {
         auth = ecdsa
         id = %any
      }
      children {
         red {
            local_ts = 192.168.99.0/24
            remote_ts = dynamic
         }
      }
   }
}
```

The `authorities` configuration now causes strongswan to fail to start due to having an explicit curve ca configured. If I switch to a different CA (rsa key)  things work fine. The  other observation is that the end entity cert ` /etc/swanctl/x509/router.crt` is signed by the explicit curve CA.  Strongswan loads that fine and doesn't complain about it even though it fails pki validation for example. I assume that is inline with not validating local certs, but the CA is also local and it gets not only rejected, but causes strongswan to not start, not sure if CAs should be treated the same when it comes local validation. Although I like the idea of always validating regardless of  local/remote config.

 I did come across: https://github.com/strongswan/strongswan/commits/reject-end-entity

That seems to enforce local cert validation but only for peers, not the local side as I have it above. Is there a way to enforce validation of certs for all sides, local and remote?

**Reply by tobiasbrunner** [reply]
> The `authorities` configuration now causes strongswan to fail to start due to having an explicit curve ca configured.

Obviously, without public key the certificate can't be loaded.

> I assume that is inline with not validating local certs

Yes, explicitly configured end-entity certificates are ultimately trusted.

> but the CA is also local and it gets not only rejected, but causes strongswan to not start

It's not about trust here, though. The key is not parsed, so the certificate file is invalid. It's just a lot easier to reject keys with explicit parameter encoding like this than adding a way to do that later during the certificate validation.

> That seems to enforce local cert validation but only for peers

It simply prevents that peers can use certificates that are locally configured (i.e. it does not force a validation if you configured the peer's cert, the connection would simply not get established).

> Is there a way to enforce validation of certs for all sides, local and remote?

No, currently not. What would be the value in this?

**Comment by Jafaral** [comment]
@tobiasbrunner , Thank you for the responses. Rejecting the key from the beginning is fine, thought it would be nice to  know/log why it was rejected in the first place; good for debugging purposes. The important piece however is these keys get rejected as they should.

>> >>Is there a way to enforce validation of certs for all sides, local and remote?

>> No, currently not. What would be the value in this?

Mostly, just to use the same systematic validation process for all certs.  If a single admin configure the box then probably it doesn't make a difference. However, if you have a system where  system-wide root of trust configuration is managed by a security admin who is different from a sys admin who configure end entity certs, then local validation become important. You only want to allow certs that are valid given the root of trust as set by the security admin.


In all cases. the  branch `1796-ecdsa-explicit` looks good and works as expected, and I suggest merging it to master/beta-six.

Looking forward for a new tag on beta-six that I can try.

**Reply by tobiasbrunner** [reply]
> ...then local validation become important. You only want to allow certs that are valid given the root of trust as set by the security admin.

What's the point? If the locally configured cert is not trusted, it will just get rejected by the peers.

**Comment by Jafaral** [comment]
The peer might be independently administered and have the signing CA cert of the local cert installed or even explicitly trusting the local cert just like how strongSwan behaves now. so the local cert might be trusted remotely for several reasons.
